lib/ui/painting/path.h

```diff
@@ -108,16 +111,39 @@ class CanvasPath : public RefCountedDartWrappable<CanvasPath> {
   bool op(CanvasPath* path1, CanvasPath* path2, int operation);
   void clone(Dart_Handle path_handle);
 
-  const SkPath& path() const { return path_; }
+  const SkPath& path() const { return tracked_path_->path_; }
 
   size_t GetAllocationSize() const override;
 
   static void RegisterNatives(tonic::DartLibraryNatives* natives);
 
+  // Called by the shell at the end of a frame after notifying Dart about idle
+  // time.
+  //
+  // This method will flip the volatility bit to false for any paths that have
+  // survived the |number_of_frames_until_non_volatile|.
+  static void updatePathVolatility();
+
+  virtual void ReleaseDartWrappableReference() const override;
+
  private:
+  struct TrackedVolatilePath {
+    bool tracking_volatility = false;
+    int frame_count = 0;
+    SkPath path_;
+  };
+
   CanvasPath();
 
-  SkPath path_;
+  static constexpr int number_of_frames_until_non_volatile = 2;
+  static std::mutex volatile_paths_mutex_;
+  static std::unordered_set<std::shared_ptr<TrackedVolatilePath>>
```

Comment:
Can this be moved to an object that lives at Shell scope and can be accessed through the UIDataState?
(similar to something like IOManager)

That would have some advantages:
* no need for a mutex if this object is only accessed on the UI thread
* if multiple engine instances exist, then one engine's NotifyIdle will not affect other engines' paths
* paths in the set will be cleaned up when the shell is deleted

Reply:
That SGTM, but I think we'd still need the mutex because the erase call could come from a Dart VM worker thread.

Reply:
Although I suppose we could just do a post task for it...

Reply:
The problem with this will be we don't know if we're in isolate scope or not when a path gets collected.

Reply:
Ok - I made this more or less work by sharing the task runner and just grabbing out the tracker when we create the path, at wihch point we'll always have isolate scope.

I think this should be correct.